recipe/2024-01-support-private-terraform-repository.md

```diff
@@ -0,0 +1,286 @@
+
+## Design
+### Design details
+Today, we support only Terraform registry and HTTP URLs as allowed module sources for Terraform recipe template paths. Git public repository are also supported by providing the HTTP URl to the module directory. But to support repositories we need a way to authenticate the git account where the modules are stored.
+
+Git provides different ways to authenticate:
+
+#### Personal Access Token:
+Users need to a Git personal access token with very limited access (just read access) and also specify token validity. And this is used along with the username to clone the terraform module repository through HTTPS as part of `terraform init`.
+
+#### SSH key:
+SSH key can be used to provide access to private git repository. But it requires adding the generated ssh key to the users account.
+
+#### Service Principal(Only for Azure DevOps Git):
+We could use the Azure Service Principal details used for Azure scope to authenticate Azure DevOps Git. But most often users have diff tenant IDs for the production environment and for git.
+
+Personal Access Token with Username way of authentication is supported by git from most of the platforms(e.g Bitbucket,Gitlab,Github etc) and generic git repository URL format as template path can be used to while registering recipe to access the terraform modules from private git repository from any platform.
+
+Generic Git Repository URL format:
+```diff
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'corerp-resources-recipe-env'
+  location: location
+  properties: {
+    ...,
+    recipes: {
+      'Applications.Datastores/mongoDatabases':{
+        recipe1: {
+          templateKind: 'terraform'
++          templatePath: "git::https://{username}:{PERSONAL_ACCESS_TOKEN}@example.com.com/test-private-repo.git"
+        }
+      }
+    }
+  }
+}
+```
+this takes latest as the default version if not provided. To specify a particular version the URL should be appended with `ref=<version>"`
+E.g:
+```
+templatePath: "git::https://{username}:{PERSONAL_ACCESS_TOKEN}@example.com.com/test-private-repo.git?ref=v1.2.0"
+```
+
+Since adding sensitive information like tokens to the terraform configuration files which may be stored in version control can pose security issues. So we could use different ways to store the github credentials.
+
+Option 1: Saving it as part of the environment resource.
+
+Add a new property `recipeConfig` write-only to store the information about the git credentials, and have a custom action `getRecipeConfiglike` (like `list-secrets`) to get these details. 
```

Comment:
I know this isn't part of the actual proposal, but I wanted to comment anyway 😆 

We also don't have to provide a custom action for everything that's a secret. Custom actions should be rare because they are complex to use and require extra work on our part to build.

I'd suggest that the right principle is to be scenario-based. We have custom actions on standard resources because they are useful for passing settings into your code.

Reply:
+1 having additional POST action is not useful unless you model recipeconfig as separate resource model.